### PR TITLE
Credit curve calibrator

### PR DESCRIPTION
**Pull request by yukiiwashita**
* Credit curve calibrators.
* 2 versions of CDS templates.
* Removed credit.cds in the pricer package to be coherent to the product package.

### CODE REVIEW

File: modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveDefinition.java [hunk 1]
```diff
@@ -79,7 +79,7 @@
    * The nodes are used to find the par rates and calibrate the curve.
    */
   @PropertyDefinition(validate = "notNull")
-  private final ImmutableList<IsdaCreditCurveNode> curveNodes;
+  private final ImmutableList<? extends IsdaCreditCurveNode> curveNodes;
```
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveDefinition.java]: We don't use `? extends` in bean fields. See `InterpolatedNodalCurveDefinition` and the `nodes` property which uses `builderType` to get this effect.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/FastCreditCurveCalibrator.java [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+            guess[i] = Math.max(minValue, guess[i]);
+            final double[] bracket =
+                BRACKETER.getBracketedPoints(func, guess[i], 1.2 * guess[i], minValue, Double.POSITIVE_INFINITY);
+            final double zeroRate = ROOTFINDER.getRoot(func, bracket[0], bracket[1]);
+            creditCurve = creditCurve.withParameter(i, zeroRate);
+          }
+          break;
+        }
+        default:
+          throw new IllegalArgumentException("unknow case " + getArbHanding());
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/FastCreditCurveCalibrator.java]: Spelling - Unknown

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/FastCreditCurveCalibrator.java [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+    private final double _accruedYearFraction;
+    private final double _productEffectiveStart;
+    private final int _startPeriodIndex;
+
+    public Pricer(ResolvedCds cds, CreditDiscountFactors yieldCurve, DoubleArray creditCurveKnots, double fractionalSpread,
+        double pointsUpfront, double lgd, LocalDate stepinDate, LocalDate effectiveStartDate, LocalDate settlementDate,
+        double accruedYearFraction) {
+
+      _accruedYearFraction = accruedYearFraction;
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/FastCreditCurveCalibrator.java]: Remove underscores and reformat to match Strata (args on new lines)

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/IsdaCompliantCreditCurveCalibrator.java [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+  }
+
+  protected IsdaCompliantCreditCurveCalibrator(AccrualOnDefaultFormula formula, ArbitrageHandling arbHandling) {
+    this.arbHandling = ArgChecker.notNull(arbHandling, "arbHandling");
+    this.formula = ArgChecker.notNull(formula, "formula");
+    this.tradePricer = new IsdaCdsTradePricer(formula);
+  }
+
+  //-------------------------------------------------------------------------
+  protected ArbitrageHandling getArbHanding() {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/IsdaCompliantCreditCurveCalibrator.java]: Protected methods need Javadoc

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/SimpleCreditCurveCalibrator.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+   * @param formula  the accrual-on-default formula
+   */
+  public SimpleCreditCurveCalibrator(AccrualOnDefaultFormula formula) {
+    super(formula);
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  NodalCurve calibrate(ResolvedCdsTrade[] calibrationCDSs, double[] premiums, double[] pointsUpfront, CurveName name,
+      LocalDate valuationDate, CreditDiscountFactors discountFactors, RecoveryRates recoveryRates, ReferenceData refData) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/SimpleCreditCurveCalibrator.java]: New lines

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/SimpleCreditCurveCalibrator.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+          : ROOTFINDER.getRoot(func, bracket[0], bracket[1]); //Negative guess handled
+      creditCurve = creditCurve.withParameter(i, zeroRate);
+    }
+
+    return creditCurve;
+  }
+
+  private Function<Double, Double> getPriceFunction(int index, ResolvedCdsTrade cds, double flactionalSpread,
+      double pointsUpfront, LocalDate valuationDate, NodalCurve creditCurve, CreditDiscountFactors discountFactors,
+      RecoveryRates recoveryRates, ReferenceData refData) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/SimpleCreditCurveCalibrator.java]: New lines

File: modules/product/src/test/java/com/opengamma/strata/product/credit/type/TenorCdsTemplateTest.java [hunk 7]
```diff
@@ -135,15 +135,15 @@ public void test_createTrade_withFee() {
...
   }
 
   public void test_serialization() {
-    CdsTemplate test = CdsTemplate.of(TENOR_10Y, CONV1);
+    TenorCdsTemplate test = TenorCdsTemplate.of(TENOR_10Y, CONV1);
     assertSerialization(test);
   }
 
-}
+}
```
Comment (jodastephen) [inline modules/product/src/test/java/com/opengamma/strata/product/credit/type/TenorCdsTemplateTest.java]: New line at end of file

File: modules/product/src/main/java/com/opengamma/strata/product/credit/type/DatesCdsTemplate.java [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+  @PropertyDefinition(validate = "notNull")
+  private final LocalDate startDate;
+  /**
+   * The end date.
+   * <p>
+   * The end date of the underling CDS product.
+   * This date can be modified following the rule in {@code convention}. 
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final LocalDate endDate;
```
Comment (jodastephen) [inline modules/product/src/main/java/com/opengamma/strata/product/credit/type/DatesCdsTemplate.java]: A "template" is supposed to be able to create the start/end date from the trade date, so this seems odd.

Comment (yukiiwashita) [inline modules/product/src/main/java/com/opengamma/strata/product/credit/type/DatesCdsTemplate.java]: Added javadoc to `TenorCdsTemplate`.

**Pull request merged**